lib/ansible/modules/files/lineinfile.py

```diff
@@ -250,10 +248,27 @@ def present(module, dest, regexp, line, insertafter, insertbefore, create,
             except Exception as e:
                 module.fail_json(msg='Error creating %s Error code: %s Error description: %s' % (b_destpath, e[0], e[1]))
 
-        b_lines = []
-    else:
-        with open(b_dest, 'rb') as f:
-            b_lines = f.readlines()
+    if module.check_mode:
+        _present_data_manipulator(module, dest, regexp, line, insertafter, insertbefore,
+                                  backup, backrefs, firstmatch)
+
+    with flock.lock_file(dest):
+        _present_data_manipulator(module, dest, regexp, line, insertafter, insertbefore,
+                                  backup, backrefs, firstmatch)
+
+
+def _present_data_manipulator(module, dest, regexp, line, insertafter, insertbefore,
+                              backup, backrefs, firstmatch):
+
+    diff = {'before': '',
+            'after': '',
+            'before_header': '%s (content)' % dest,
+            'after_header': '%s (content)' % dest}
+
+    b_dest = to_bytes(dest, errors='surrogate_or_strict')
+
+    with open(b_dest, 'rb') as f:
+        b_lines = f.readlines()
```

Comment:
Looking good.  I'd do a little more reorganization of the content, though.

* Move these lines for reading from the file into ```present()```
* Move the ```write_changes()``` call into ```present()```.
* Move the module.exit_json() into ```present()``` (or higher, into ```main()```, but I can see that it  was in present() before so it's a preexisting problem).
* Change what we're passing in to ```_present_data_manipulator``` so that we do not pass in things that have to do with the file (module, dest, and backup), but it does include b_lines.
* Assembling the diff results and checking file attributes probably gets moved out to the higher level as well, although I'm not entirely happy with that.... those seem to float in the middle of the overall organization that we're shooting for.
* (For a separate PR) we can also change ```_present_data_manipulator()``` to return a copy of b_lines instead of modifying the data in place.  In place uses less memory but a function without side effects is generally better code. (As an example, think about the additional flexibility of python's ```sorted()``` function over ```$iterable.sort()```

The object is to make this helper function be only about manipulating the data and returning the new data.  Everything about the file itself or the fact that we're part of an Ansible module rather than some other script is removed from the helper.  This is a better organization for people to read in the future as all of the inputs and all of the outputs of ```_present_data_manipulator``` then become obvious to someone who is reading the code.

One way to judge this is to think of how you'd unit test this new function.  If you succeed in moving everything about file handling and module outside of the new function, then the new function is very straightforward to unittest.  You just pass in different values of b_lines (the data to be manipulated) and the options which change behaviour, regexp, line, insertbefore, insertafter, backrefs, and firstmatch, and then you see whether the lines that are output are what you expected or not.

By contrast, if the file manipulation and module is part of the new code, then you end up having to mock file objects and AnsibleModule in order to unit test the code.